Blog - Querying Relational Databases With SQLAlchemy in Python (#435)
earthly/website: blog/_posts/2023-05-18-query-database-sqlalchemy-python.md

- **Data Integrity and Transactions**: SQLAlchemy provides support for transactions and data integrity constraints, such as foreign keys, unique constraints, and check constraints. This helps to ensure that data is consistent and accurate across different tables in the database.
- **Cross-database Compatibility:** SQLAlchemy provides a consistent API for interacting with different database systems, making it easy to write database applications that can work with different databases.

## Setting Up Sqlalchemy

Before getting started with SQLAlchemy, you need to first install the SQLAlchemy library. Execute the following command in your terminal or command prompt. This command will install the SQLAlchemy library via [pip](https://pypi.org/project/pip/):

~~~{.bash caption=">_"}
pip install SQLAlchemy
~~~

If you have it installed successfully, you should have the following output:

<div class="wide">
![Installing the sqlalchemy library]({{site.images}}{{page.slug}}/oqHbE7L.png)
</div>

> Create a file named `[main.py](http://main.py)` (you can of course name this file whatever you want) and import the following from the SQLAlchemy library:

Reviewer: ```suggestion
Create a file named `main.py` (you can of course name this file whatever you want) and import the following from the SQLAlchemy library:
```

- **`ForeignKey`**: This class provides a way to define foreign keys in SQLAlchemy classes. Foreign keys are used to establish relationships between tables in a relational database.
- **`Column`**: This class is used to define columns in SQLAlchemy classes. Columns represent individual fields in a database table.
- **`String`**: This class is used to define a string data type for SQLAlchemy columns.
- **`Integer`**: This class is used to define an integer data type for SQLAlchemy columns.
- **`CHAR`**: This class represents a character column in a database table.
- **`CheckConstraint`**: This class defines a check constraint in a table, which is a condition that must be met for the data to be valid.
- **`join`**: This function constructs a SQL join between two or more tables, which allows data from different tables to be combined in a single query result.

Next, add the following line of code to import the **`declarative_base`** function from the **`sqlalchemy.ext.declarative`** module.

~~~{.python caption="main.py"}
from sqlalchemy.ext.declarative import declarative_base
~~~

> <aside>
> 💡 The **`declarative_base`** function is used to create a base class that is used to define the database schema. This base class allows us to define the database schema in a way that is more like defining regular Python classes, and it makes it easy to create new database tables and columns.
> 
> After creating the base class by using the **`declarative_base`** function, we can define classes that inherit from the base class, and each of these classes represents a table in the database. We can define the columns of the table as attributes of the class, and we can define relationships between tables using attributes as well. Once we have defined all of our classes, we can use SQLAlchemy to create the database schema and interact with the database using the classes we have defined.
> 
> </aside>

Reviewer: ```suggestion
<div class="notice--big--primary">

💡 The **`declarative_base`** function is used to create a base class that is used to define the database schema. This base class allows us to define the database schema in a way that is more like defining regular Python classes, and it makes it easy to create new database tables and columns.

After creating the base class by using the **`declarative_base`** function, we can define classes that inherit from the base class, and each of these classes represents a table in the database. We can define the columns of the table as attributes of the class, and we can define relationships between tables using attributes as well. Once we have defined all of our classes, we can use SQLAlchemy to create the database schema and interact with the database using the classes we have defined.

</div>
```

~~~{.python caption="main.py"}
from sqlalchemy.orm import sessionmaker, relationship
~~~

These classes do the following;

1. **`sessionmaker`**: It is a factory class that returns a new **`Session`** class when called with a database engine. The **`Session`** class is used to create database sessions, which are used to communicate with the database and perform CRUD (Create, Read, Update, Delete) operations.
2. **`relationship`**: It is a function that is used to define relationships between two database tables. It is used to specify how data in one table is related to data in another table. The **`relationship`** function is used to create a bidirectional relationship between two database tables.

Finally, create a new instance of the **`declarative_base`** class provided by SQLAlchemy and assign it to the variable **`Base`**. This instance will be used as a base class for the definition of database models using the SQLAlchemy ORM (Object-Relational Mapping) framework.

~~~{.python caption="main.py"}
Base = declarative_base()
~~~

## Creating a Table

Now that we have SQLAlchemy all setup, let's see how we can create a table and store it in an SQLite database.

> First, we'll define an SQLAlchemy model class **`User`** that represents a table called *users* in a database, this could be any class of your choice (a **`Persons`** class or a **`Students`** class):

Reviewer: ```suggestion
First, we'll define an SQLAlchemy model class **`User`** that represents a table called *users* in a database. This could be any class of your choice (a **`Persons`** class or a **`Students`** class):
```

2023-03-27 11:59:29,853 INFO sqlalchemy.engine.Engine 
CREATE TABLE pets (
        "ID" INTEGER NOT NULL, 
        "NAME" VARCHAR, 
        "OWNER" INTEGER, 
        PRIMARY KEY ("ID"), 
        FOREIGN KEY("OWNER") REFERENCES users (ssn)
)

2023-03-27 11:59:29,853 INFO sqlalchemy.engine.Engine [no key 0.00017s] ()
2023-03-27 11:59:29,871 INFO sqlalchemy.engine.Engine COMMIT
2023-03-27 11:59:29,879 INFO sqlalchemy.engine.Engine BEGIN (implicit)
2023-03-27 11:59:29,881 INFO sqlalchemy.engine.Engine INSERT INTO pets \
("ID", "NAME", "OWNER") VALUES (?, ?, ?)
2023-03-27 11:59:29,881 INFO sqlalchemy.engine.Engine \
[generated in 0.00031s] [(1, 'Dog', 1000), (2, 'Cat', 1000), \
(3, 'Rabbit', 1003), (4, 'Rabbit', 1002)]
2023-03-27 11:59:29,884 INFO sqlalchemy.engine.Engine COMMIT
~~~

Now if you view your database file, which is, in this case, called *mydb.db* you should have two tables, select the `pets` table and you should have the below output:

<div class="wide">
![Viewing the pets table]({{site.images}}{{page.slug}}/wXxuGXQ.png)
</div>

## Querying a Table

Now that you have seen how to create relationships between tables, how about querying a table with SQLAlchemy?

> Up until now, you are expected to have the following snippets in your *main.py*  file:

Reviewer: ```suggestion
Up until now, you are expected to have the following snippets in your `main.py`  file:
```

~~~{.python caption="main.py"}
output = session.query(Pet).filter(User.lastname == "Doe")
for i in output:
    print(i)
~~~

<div class="wide">
![Filtering out users with last name (doe) from users table]({{site.images}}{{page.slug}}/FoESeM4.png)
</div>

From the image above, we have two results, **`John Doe`** and **Jane Doe`**.

Additionally, you can also search for all the pets in the database that have the name *Rabbit* using the below line of code:

~~~{.python caption="main.py"}
output = session.query(Pet).filter(Pet.name == "Rabbit")
for i in output:
    print(i)
~~~

The code above searches for all the pets in the database that have the name *Rabbit* using the [**`filter()`**](https://docs.sqlalchemy.org/en/20/orm/queryguide/query.html#sqlalchemy.orm.Query.filter) method. Then, it loops through the results and prints each pet's information to the console using a **`for`** loop and the **`print()`**function.

<div class="wide">
![Filtering all pets with name (Rabbit) from pets table]({{site.images}}{{page.slug}}/HjsCVKo.png)
</div>

From the output above you can see that you have two results, a pet named *Rabbit* for users with *ssn* numbers 1003 and 1002 respectively.

> Finally, you can retrieve all the users whose country starts with the letter "M" from the ***`users`** table using the following command:

Reviewer: ```suggestion
Finally, you can retrieve all the users whose country starts with the letter "M" from the `users` table using the following command:
```
